PYME/IO/DataSources/BioformatsDataSource.py

```diff
@@ -109,7 +110,24 @@ def __init__(self, filename, taskQueue=None, chanNum = 0):
         
         self.additionalDims = 'TC'
                 
+    def setSeries(self, series):
```

Comment:
What is the use case for calling `setSeries()` on an existing data source? In general most consuming code (dh5view, analysis, etc ...) expect a `DataSource` to be immutable (or at the very least, have constant dimensionality). I suspect that it's not going to be a good idea to to change the series that a data source points to on the fly.

Reply:
I wanted to be able to loop through the different images in the series with a single load call. I figured it would be more efficient. We can drop it if you feel it's unsafe. I have no problem loading the same file 5 times to get 5 different series within it.

Reply:
I think it's preferable to create 5 different DataSources. That said, we need a way of working out what is in the file, and ultimately selecting which series we want in the UI (or on the command line for recipes etc ...).

One way of addressing this would be to:
- change `BioformatsDataSource` so that the filename parameter (potentially renamed to file - would need to check how it's called) could take either a filename or a `bioformats.ImageReader` instance (or potentially our own proxy around a bioformats file so we can track the filename etc ...).
- In `image.py` (or wherever we call it) instantiate the image reader, and query the number of series. If we have a GUI and there is more than one series, prompt the user to see which series they want to open. Pass the constructed `ImageReader` that we constructed to the data source.


That leaves us with the command line - which we could potentially defer to later - I'd be in favour of using the http URL query syntax e.g. `/path/to/somefile.lif?series=xxx` in the filename.

Reply:
Doing it this way I can see a clear path to having something like the UI that FIJI has for loading multi-series files - where you see a list of series with their dimensionality, series name (if we can get it) and (potentially - if there is sufficient demand) a thumbnail.

Reply:
Yes, I agree. I'll have to look a little more into the http URL query syntax. Series name is accessible through `self.bff.rdr.getSeriesMetadata(<series number>)`, I think. Maybe the `BioformatsDataSource` should contain the whole LIF (or whatever bioformats image) and then if we want to open a series from it we just spin off an `ArrayDataSource` containing the series?

Reply:
You don't want to load everything and then spin out an `ArrayDataSource` as it then forces everything into memory (a problem if have large data sets). At present, we load lazily, meaning that we only have a few frames in memory at once.

Reply:
Oh, that's right.

OK, so, do we want almost exactly what we have here then? We probably want to put the option of which series to load in the GUI, but the data source should just accept a series as an argument and just load that one and never change, right?

Reply:
I looked into the URL format. I don't see a clear reason to go with it over just loading the file and making a getSeries() call. Seems to be just variations on a pointer. Is there more to it?

Reply:
We'd still be loading a file and calling getSeries(), my idea with the URL format would just be a way of specifying which series we want on the command line (for cases where we don't want to prompt the user and/or don't have a GUI - e.g. automated analysis in bakeshop). This is why I said we could defer it. We currently use something similar with the `SupertileDataSource`.

Reply:
We're a bit out of sync - replying to your comment 2nd to last comment: yep - pretty much what we want. I'd ideally:
- keep the series argument to the constructor
- back out the `getSeries()`, `setSeries()`, `setXYZTX()` calls (which are potentially confusing if we never expect them to get called)
- remove the `nSeries` property
- remove the current check on the number of series in the constructor.
- add the following to the constructor (before setting x and y sizes etc):
```python
if isinstance(filename, BioformatsFile): #TODO - change filename parameter name
    self.bff = filename.bff
else:
    self.bff = bioformats.ImageReader(filename)

if not series is None:
    self.bff.rdr.setSeries(series)
```
I'd then copy the `getSeries` and `nSeries` logic to a new class:

```python
class BioformatsFile(object):
    def __init__ (self, filename):
        self.filename = filename
        self.bff = bioformats.ImageReader(filename)

    def get_series_info(self):
        # potentially rename
        # return a list of series names in the file, potentially also with dimensionality etc ...
```

Usage in `Image._loadBioformats()` 

```python
series_num = None
#try parsing filename for series URL parameter, over-write series_num if sucessful

bioformats_file = BioformatsDataSource.BioformatsFile(filename)
series_info = bioformats_file.get_series_info()

if series_num is None and len(series_info) >1:
    print("File has multiple series, need to pick one")

    if self.haveGUI:
        #wx multichoice dialog on series_info list, set series_num to result
    else:
        logger.warning('No GUI, using 0th series')

BioformatsDataSource.DataSource(bioformats_file, series=series_num)
```

Reply:
some of the ordering above is wonky - create the `BioformatsFile` class and copy logic before removing from the data source.

Reply:
Actually thinking about it a bit more, the `BioformatsFile` class as such a thin proxy is a bit gross - it should probably be a subclass of `ImageReader` rather than a class which simply holds a reference to an `ImageReader` instance. We're basically just adding an easy way to get the number of series and present them in a GUI.

Reply:
Ah, that's true. So subclass `ImageReader` and add the standard `DataSource` functions like `getSlice()`, `getSliceShape()`, etc.? Or multi-inherit `BaseDataSource` and `ImageReader`?

Reply:
Neither - Subclass `ImageReader` to provide simpler access to info about the different series (no additional functions like `getSlice()`) - this is just to make the UI for series selection easier - you could in principle put all this logic in `Image` and make it call `bff.rdr.getSeriesCount()` and `[bff.rdr.getSeriesMetadata(n) for n in range(bff.rdr.getSeriesCount())], but it might be a little cleaner to keep it out of there. 

Make the `DataSource` own a reference to an `ImageReader` (or subclass thereof) (as it currently does).

Reply:
Quick correction on the dialog in the GUI - you want a `wx.SingleChoiceDialog`

Reply:
Oh I see. I missed `BioformatsFile`.